From 35efc1d1999eb7a02817426ebda51c3cf8ef4cbc Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 19 Nov 2025 18:27:18 +0100
Subject: [PATCH] irqwork: Move data struct to a types header

... to avoid header recursion hell.

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Reviewed-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Link: https://patch.msgid.link/20251119172550.152813625@linutronix.de
Backport-note: drop struct rcuwait irqwait field; 5.15 struct irq_work has no irqwait and its irq_work_sync() does not use rcuwait. Keeping it pulled in rcuwait.h and caused an incomplete-type build failure via a circular include (rcuwait.h -> rcupdate.h -> srcu -> srcutiny.h -> irq_work_types.h).
---
 include/linux/irq_work.h       |  6 +-----
 include/linux/irq_work_types.h | 13 +++++++++++++
 2 files changed, 14 insertions(+), 5 deletions(-)
 create mode 100644 include/linux/irq_work_types.h

diff --git a/include/linux/irq_work.h b/include/linux/irq_work.h
index ee5f9120c4d73..633cc55b99798 100644
--- a/include/linux/irq_work.h
+++ b/include/linux/irq_work.h
@@ -2,6 +2,7 @@
 #ifndef _LINUX_IRQ_WORK_H
 #define _LINUX_IRQ_WORK_H
 
+#include <linux/irq_work_types.h>
 #include <linux/smp_types.h>
 
 /*
@@ -13,11 +14,6 @@
  * busy      NULL, 2 -> {free, claimed} : callback in progress, can be claimed
  */
 
-struct irq_work {
-	struct __call_single_node node;
-	void (*func)(struct irq_work *);
-};
-
 #define __IRQ_WORK_INIT(_func, _flags) (struct irq_work){	\
 	.node = { .u_flags = (_flags), },			\
 	.func = (_func),					\
diff --git a/include/linux/irq_work_types.h b/include/linux/irq_work_types.h
new file mode 100644
index 0000000000000..0b92472c771c4
--- /dev/null
+++ b/include/linux/irq_work_types.h
@@ -0,0 +1,13 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef _LINUX_IRQ_WORK_TYPES_H
+#define _LINUX_IRQ_WORK_TYPES_H
+
+#include <linux/smp_types.h>
+#include <linux/types.h>
+
+struct irq_work {
+	struct __call_single_node	node;
+	void				(*func)(struct irq_work *);
+};
+
+#endif
-- 
2.43.0

